ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoCommandDeduplicationSpec.scala

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.dao
+
+import java.time.Instant
+import java.util.UUID
+
+import com.daml.ledger.api.domain.CommandId
+import com.daml.ledger.participant.state.index.v2.{
+  CommandDeduplicationDuplicate,
+  CommandDeduplicationNew
+}
+import org.scalatest.{AsyncFlatSpec, Matchers}
+
+private[dao] trait JdbcLedgerDaoCommandDeduplicationSpec {
+  this: AsyncFlatSpec with Matchers with JdbcLedgerDaoSuite =>
+
+  behavior of "JdbcLedgerDao (comman deduplication)"
+
+  it should "correctly deduplicate a command" in {
+    val commandId: CommandId = CommandId(UUID.randomUUID.toString)
+    for {
+      original <- ledgerDao.deduplicateCommand(commandId, List(alice), t(0), t(5000))
+      duplicate <- ledgerDao.deduplicateCommand(commandId, List(alice), t(500), t(5500))
+    } yield {
+      original shouldBe CommandDeduplicationNew
+      duplicate shouldBe CommandDeduplicationDuplicate(t(5000))
+    }
+  }
+
+  it should "correctly deduplicate commands with multiple submitters" in {
+    val commandId: CommandId = CommandId(UUID.randomUUID.toString)
+    for {
+      original <- ledgerDao.deduplicateCommand(commandId, List(alice, bob), t(0), t(5000))
+      duplicate1 <- ledgerDao.deduplicateCommand(commandId, List(alice, bob), t(500), t(5500))
+      duplicate2 <- ledgerDao.deduplicateCommand(commandId, List(bob, alice), t(500), t(5500))
+      duplicate3 <- ledgerDao.deduplicateCommand(
+        commandId,
+        List(alice, bob, alice),
+        t(500),
+        t(5500))
+    } yield {
+      original shouldBe CommandDeduplicationNew
+      duplicate1 shouldBe CommandDeduplicationDuplicate(t(5000))
+      duplicate2 shouldBe CommandDeduplicationDuplicate(t(5000))
+      duplicate3 shouldBe CommandDeduplicationDuplicate(t(5000))
+    }
+  }
+
+  it should "not deduplicate a command after it expired" in {
+    val commandId: CommandId = CommandId(UUID.randomUUID.toString)
+    for {
+      original1 <- ledgerDao.deduplicateCommand(commandId, List(alice), t(0), t(100))
+      original2 <- ledgerDao.deduplicateCommand(commandId, List(alice), t(101), t(200))
+    } yield {
+      original1 shouldBe CommandDeduplicationNew
+      original2 shouldBe CommandDeduplicationNew
+    }
+  }
+
+  it should "not deduplicate a command after it was cancelled" in {
```

Comment:
```suggestion
  it should "not deduplicate a command after its deduplication was stopped" in {
```

I'm suggesting this because it looks to me like you're testing a lower-level behavior here (that is used by an higher-level one, i.e. command cancellation) but feel free to ignore if that's not the case.